internal/graph/singleflight_resolver_test.go

```diff
@@ -18,156 +17,146 @@ import (
+			StoreID:            storeID,
+			TupleKey:           tuple.NewTupleKey("doc:2", "a4", "user:jon"),
+			ContextualTuples:   nil,
+			ResolutionMetadata: &ResolutionMetadata{Depth: defaultResolveNodeLimit},
+		})
+		require.NoError(t, err)
+		require.False(t, resp.GetAllowed())
+	})
 
-	checker := NewLocalChecker(
-		storagewrappers.NewCombinedTupleReader(ds, []*openfgav1.TupleKey{}),
-		WithSingleflightResolver(),
-	)
-	t.Cleanup(checker.Close)
-
-	resp, err := checker.ResolveCheck(ctx, &ResolveCheckRequest{
-		StoreID:            storeID,
-		TupleKey:           tuple.NewTupleKey("document:1", "viewer3", "user:jon"),
-		ContextualTuples:   nil,
-		ResolutionMetadata: &ResolutionMetadata{Depth: defaultResolveNodeLimit},
+	t.Run("Check with singleflight resolver reduces DB query count when compared to not using it", func(t *testing.T) {
+		ds := memory.New()
+		t.Cleanup(ds.Close)
+		err := ds.Write(context.Background(), storeID, nil, []*openfgav1.TupleKey{
+			tuple.NewTupleKey("doc:2", "a1", "user:jon"),
+		})
+		require.NoError(t, err)
+
+		checkerWithSingleflight := NewLocalChecker(
+			storagewrappers.NewCombinedTupleReader(ds, []*openfgav1.TupleKey{}),
+			WithSingleflightResolver(),
+		)
+		t.Cleanup(checkerWithSingleflight.Close)
+
+		// The results of the singleflight resolver are not deterministic.
+		// For better test reliability, the test is repeated a number of times
+		// and then assertions made within a reasonable threshold.
+		testIterations := 3
+
+		var dbReadsWith uint32
+		for i := 0; i < testIterations; i++ {
+			resp, err := checkerWithSingleflight.ResolveCheck(ctx, &ResolveCheckRequest{
+				StoreID:            storeID,
+				TupleKey:           tuple.NewTupleKey("doc:1", "a4", "user:jon"),
+				ContextualTuples:   nil,
+				ResolutionMetadata: &ResolutionMetadata{Depth: defaultResolveNodeLimit},
+			})
+			require.NoError(t, err)
+			require.False(t, resp.GetAllowed())
+			dbReadsWith += resp.GetResolutionMetadata().DatastoreQueryCount
+		}
+
+		checkerWithoutSingleflight := NewLocalChecker(
+			storagewrappers.NewCombinedTupleReader(ds, []*openfgav1.TupleKey{}),
+		)
+
+		var dbReadsWithout uint32
+		for i := 0; i < testIterations; i++ {
+			resp, err := checkerWithoutSingleflight.ResolveCheck(ctx, &ResolveCheckRequest{
+				StoreID:            storeID,
+				TupleKey:           tuple.NewTupleKey("doc:1", "a4", "user:jon"),
+				ContextualTuples:   nil,
+				ResolutionMetadata: &ResolutionMetadata{Depth: defaultResolveNodeLimit},
+			})
+			require.NoError(t, err)
+			require.False(t, resp.GetAllowed())
+			dbReadsWithout += resp.GetResolutionMetadata().DatastoreQueryCount
+		}
+
+		require.Less(t, dbReadsWith, dbReadsWithout) //singleflight resolver will always result in fewer DB reads than without
+
+		require.LessOrEqual(t, dbReadsWith, uint32(testIterations+2)) // A buffer of two DB reads to ensure test reliability
+		require.GreaterOrEqual(t, dbReadsWith, uint32(testIterations))
```

Comment:
You are comparing the outcome of the sum of 2 experiments that ran N times each. I think this is mathematically equivalent to looking at the outcome of each individual experiment, but I feel like the latter is easier to read. I suggest moving the assertions to within each loop. 

```go
// for each iteration of singleflight
require.GreaterOrEqual(t, res.ResolutionMetadata.DatastoreQueryCount, 1)
require.LessOrEqual(t, res.ResolutionMetadata.DatastoreQueryCount, 2)
```

`testIterations := 3` i also recommend making this larger to increase the chance that you'll see both results. say 1000

Reply:
The problem here is that 99% is roughly equivalent to 100%. That's the whole point here, the tests aren't deterministic. 

Taking the sum of several iterations is more likely to result in a higher pass-rate because the aggregation will regress to the mean. Your suggestion of running 1000 iterations and testing the result of each assertion is the opposite, it will fail if 1/1000 trials fails, which is an almost certainty.

Reply:
Sorry i'm confused,

> Your suggestion of running 1000 iterations and testing the result of each assertion is the opposite, it will fail if 1/1000 trials fails,

but why would it fail? If the run for singleflight isn't 1 or 2, then that is a legit bug and we should know about it

Reply:
> If the run for singleflight isn't 1 or 2

Evaluating the DB query count at face value doesn't prove that it works. If a single test requires two DB reads, so what? That's the same as not using it. Granted, we could add that assertion to compliment the existing assertions, but it would only prove that there aren't more than two DB queries for a single check.

The test is much stronger and reliable when comparing singleflight resolver usage vs not. And the only reliable way of testing this is over multiple iterations because the results are probabilistic.